duckdb-vortex/src/vortex_extension.cpp

```diff
@@ -52,12 +55,30 @@ struct VortexScanLocalState : public LocalTableFunctionState {
 	idx_t current_row;
 	bool finished;
 	Array *array;
+	uint32_t thread_id;
+
+	explicit VortexScanLocalState(uint32_t thread_id)
+	    : current_row(0), finished(false), array(nullptr), thread_id(thread_id) {
+	}
 };
 
-struct VortexScanGlobalState : public GlobalTableFunctionState {
+struct ThreadSlot {
+	std::mutex slot_lock;
 	ArrayStream *array_stream;
-	std::mutex stream_lock;
-	bool finished;
+};
+
+struct VortexScanGlobalState : public GlobalTableFunctionState {
+	// Must be <= MAX_THREAD_COUNT.
+	std::atomic_uint32_t thread_id_counter;
+	std::atomic_bool finished;
+
+	// each thread owns a thread slot and is the only one allowed to modify the slot itself,
+	// however other threads can work-steal array batches from the slot, but taking out the mutex in the ThreadSlot
+	// Allocate MAX_THREAD_COUNT threads, the max number threads allowed by this extension.
+	duckdb::vector<unique_ptr<ThreadSlot>> file_slots;
```

Comment:
We don't need to heap allocate the `ThreadSlot`s, right?

=>

```
duckdb::vector<ThreadSlot> file_slots
```

Reply:
When accessing the vector, elements would need to be accessed by reference though. Or at least in some pattern which does not try to copy std::mutex by value.

Reply:
You can then default construct the file_slots with the reserve count directly:

```
	explicit VortexScanGlobalState()
	    : thread_id_counter(0), finished(false), file_slots(MAX_THREAD_COUNT), next_file(0),
	      filter(nullptr) {
	}
```

Reply:
Went with an array, maybe you can PR later if we want a dynamically size file slot